File: server/src/chains/rag-chain.js
import {
  RunnableSequence,
  RunnablePassthrough,
} from "@langchain/core/runnables";
import { StringOutputParser } from "@langchain/core/output_parsers";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { PGVectorStore } from "@langchain/community/vectorstores/pgvector";
import { createModel } from "../models/deepseek.js";
import { embeddings } from "../models/embedding.js";
import { pool } from "../db/postgres.js";
import { rerankDocuments } from "./reranker.js";

const PG_CONFIG = {
  pool,
  tableName: "knowledge_embeddings",
  columns: {
    idColumnName: "id",
    vectorColumnName: "embedding",
    contentColumnName: "content",
    metadataColumnName: "metadata",
  },
};

// 初始化 VectorStore（模块加载时执行一次）
const vectorStore = await PGVectorStore.initialize(embeddings, PG_CONFIG);

// 标准检索器（k=4，供原有端点兼容使用）
export const retriever = vectorStore.asRetriever({ k: 4 });

// 粗检索器（k=10，第一阶段检索）
const coarseRetriever = vectorStore.asRetriever({ k: 10 });

/**
 * 两阶段检索 — 粗检索 + LLM 重排序
 * 第一阶段：向量相似度检索 top-10
 * 第二阶段：LLM 重排序取 top-4
 * @param {string} question
 * @param {number} topK - 最终返回数量，默认 4
 * @returns {Promise<Array>} 重排序后的文档数组
 */
export async function retrieveWithRerank(question, topK = 4) {
  const docs = await coarseRetriever.invoke(question);
  return await rerankDocuments(question, docs, topK);
}

// T10: 导出 ragPrompt 供 /stream 端点使用
export const ragPrompt = ChatPromptTemplate.fromMessages([
  [
    "system",
    `你是极速购电商平台的专业客服助手小购。

请根据以下知识库内容回答用户的问题。
如果知识库中没有相关内容，请如实告知用户，不要编造信息。
回答语气友好，称呼用户为"亲"，回复简洁清晰。

知识库内容：
{context}`,
  ],
  ["human", "{question}"],
]);

// T10: 导出 formatDocs 供 /stream 端点使用
export const formatDocs = (docs) =>
  docs.map((doc) => doc.pageContent).join("\n\n---\n\n");

// 从 chunk 内容中提取最有意义的 Markdown 标题作为章节标签
function extractSectionTitle(content = "") {
  const lines = content
    .split("\n")
    .map((l) => l.trim())
    .filter(Boolean);
  const headings = [];

  for (const line of lines) {
    const match = line.match(/^(#{1,6})\s+(.+)$/);
    if (match) {
      headings.push({
        level: match[1].length,
        title: match[2].trim(),
      });
    }
  }

  if (headings.length === 0) {
    const preview = content.replace(/\s+/g, " ").trim();
    return preview.length > 10
      ? preview.slice(0, 10) + "..."
      : preview || "相关片段";
  }

  // 如果只有一个标题，直接用它
  if (headings.length === 1) return headings[0].title;

  // 优先使用二级或三级标题（比文档标题更具体的章节）
  const specific = headings.find((h) => h.level >= 2 && h.level <= 3);
  if (specific) return specific.title;

  return headings[0].title;
}

// 导出提取来源信息的工具函数
export const extractSources = (docs) =>
  docs.map((doc, idx) => ({
    content: doc.pageContent.slice(0, 120) + "...",
    source: doc.metadata.source || "未知文档",
    section: extractSectionTitle(doc.pageContent),
    page: doc.metadata.page,
    score: doc.metadata.score ?? 1 - idx * 0.05, // 降序默认分数
  }));

const model = createModel({ temperature: 0 });

// 流式模型（用于 /stream 端点的逐字输出）
export const streamingRagModel = createModel({
  temperature: 0,
  streaming: true,
});

// 流式 RAG Chain（prompt → streamingModel → parser）
export const ragStreamChain = ragPrompt
  .pipe(streamingRagModel)
  .pipe(new StringOutputParser());

// 标准 RAG Chain（原有，使用 k=4 直接检索）
export const ragChain = RunnableSequence.from([
  {
    context: (input) => retriever.pipe(formatDocs).invoke(input.question),
    question: (input) => input.question,
  },
  ragPrompt,
  model,
  new StringOutputParser(),
]);

// 重排序 RAG Chain（使用两阶段检索）
export const ragChainWithRerank = RunnableSequence.from([
  {
    context: async (input) => {
      const docs = await retrieveWithRerank(input.question);
      return formatDocs(docs);
    },
    question: (input) => input.question,
  },
  ragPrompt,
  model,
  new StringOutputParser(),
]);

// 带来源信息的 RAG Chain
export const ragChainWithSources = RunnableSequence.from([
  RunnablePassthrough.assign({
    docs: (input) => retriever.invoke(input.question),
  }),
  {
    answer: RunnableSequence.from([
      (input) => ({
        context: formatDocs(input.docs),
        question: input.question,
      }),
      ragPrompt,
      model,
      new StringOutputParser(),
    ]),
    sources: (input) => extractSources(input.docs),
  },
]);

// 重排序 + 来源信息的 RAG Chain（供 Graph 的 ragNode 使用）
export const ragChainWithSourcesAndRerank = RunnableSequence.from([
  RunnablePassthrough.assign({
    docs: async (input) => retrieveWithRerank(input.question, 4),
  }),
  {
    answer: RunnableSequence.from([
      (input) => ({
        context: formatDocs(input.docs),
        question: input.question,
      }),
      ragPrompt,
      model,
      new StringOutputParser(),
    ]),
    sources: (input) => extractSources(input.docs),
  },
]);

File: backend_py/chains/rag_chain.py
"""
RAG 检索增强链路
对照 JS: RunnableSequence.from([...])
Python:  使用 | 管道符连接，配合 RunnableParallel 并行执行
"""
import os
from dotenv import load_dotenv
from langchain_core.runnables import RunnablePassthrough
from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_community.vectorstores import PGVector
from models.deepseek import create_model
from models.embedding import embeddings

load_dotenv()

# 构建连接字符串
PG_USER = os.getenv("PG_USER", "postgres")
PG_PASSWORD = os.getenv("PG_PASSWORD", "postgres")
PG_HOST = os.getenv("PG_HOST", "localhost")
PG_PORT = os.getenv("PG_PORT", "5432")
PG_DATABASE = os.getenv("PG_DATABASE", "langchain_course")
CONNECTION_STRING = f"postgresql://{PG_USER}:{PG_PASSWORD}@{PG_HOST}:{PG_PORT}/{PG_DATABASE}"

# 1. 初始化向量存储（模块加载时执行一次）
vector_store = PGVector(
    connection_string=CONNECTION_STRING,
    embedding_function=embeddings,
    collection_name="knowledge",
)

# 每次检索返回最相似的 4 个片段
retriever = vector_store.as_retriever(search_kwargs={"k": 4})

# 2. RAG Prompt
rag_prompt = ChatPromptTemplate.from_messages([
    ("system", """你是极速购电商平台的专业客服助手小购。

    请根据以下知识库内容回答用户的问题。
    如果知识库中没有相关内容，请如实告知用户，不要编造信息。
    回答语气友好，称呼用户为"亲"，回复简洁清晰。

    知识库内容：
    {context}"""),
    ("human", "{question}"),
])


# 3. 格式化检索文档
def format_docs(docs):
    """把检索到的文档列表格式化成字符串"""
    return "\n\n---\n\n".join(doc.page_content for doc in docs)


# 4. 组装 RAG Chain
# 对照 JS:
#   RunnableSequence.from([{
#     context: (input) => retriever.pipe(formatDocs).invoke(input.question),
#     question: (input) => input.question,
#   }, ragPrompt, model, parser])
#
# Python 用字典 + 管道符实现同样的并行分支：
model = create_model(temperature=0)

rag_chain = (
    {
        "context": (lambda x: x["question"]) | retriever | format_docs,
        "question": (lambda x: x["question"]),
    }
    | rag_prompt
    | model
    | StrOutputParser()
)


# 5. 带来源信息的版本
# 对照 JS: ragChainWithSources
async def rag_chain_with_sources_invoke(question: str) -> dict:
    """
    返回 { answer, sources } 的字典
    对照 JS 的 ragChainWithSources

    关键改进：sources 按文件名去重
        - PGVector 检索时，长文档被切分成多个 chunks 入库
        - 向量空间相近的多个 chunk 会被 retriever 一起召回
        - 在 sources 列表中按文件名去重，每个文件只保留最相关的那条
        - 前端 <span class="source-tag"> 也就不会再显示重复文件名了
    """
    docs = await retriever.ainvoke(question)
    context = format_docs(docs)
    answer = await (rag_prompt | model | StrOutputParser()).ainvoke({
        "context": context,
        "question": question,
    })

    # 按文件名去重（保留第一条 = 相似度最高的片段）
    sources = []
    seen = set()
    for doc in docs:
        src = doc.metadata.get("source", "unknown")
        if src not in seen:
            sources.append({
                "content": doc.page_content[:100] + "...",
                "source": src,
            })
            seen.add(src)

    return {"answer": answer, "sources": sources}
